dali/kernels/signal/downmixing.h

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_DOWNMIXING_H_
+#define DALI_KERNELS_SIGNAL_DOWNMIXING_H_
+
+#include <cassert>
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/span.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+/**
+ * Downmix data to a single channel.
+ *
+ * @param out               output buffer (single channel)
+ * @param in                input buffer (interleaved multiple channels)
+ * @param num_samples       number of samples in each channel
+ * @param channels          number of input channels
+ * @param weights           weights used for downmixing
+ * @param normalize_weights if true, the weights are normalized so their sum is 1
+ * @tparam Out output sample type - if integral, the intermediate floating point representation
+ *         is stretched so that 0..1 or -1..1 range occupies the whole Out range.
+ * @tparam In input sample type - if integral, it's normalized to 0..1 or -1..1 range
+ * @tparam static_channels compile-time number of channels
+ */
+template <int static_channels = -1, typename Out, typename In>
+void DownmixChannels(
+    Out *out, const In *in, int64_t samples, int channels,
+    const float *weights, bool normalize_weights = false) {
```

Comment:
Not if you want to vectorize the loop.

Reply:
Can we add a `restricted` qualifier to `in` and `out`?

Reply:
This is not a C++ qualifier, unfortunately. The __restrict keyword is in C and in CUDA.

Reply:
GCC and Clang supports `__restrict__`, do you think we can use it or we want to stick to the standard?

Reply:
> Not if you want to vectorize the loop.

Then is `Span_DefaultWeights` incorrect? It does downmixing in-place

Reply:
I'm not adding it here, but I will in resampling which is utterly impossible to do in-place.

Reply:
OK, after a thought I'll add the in-place info - obviously with the caveat that Out and In should be of the same time - otherwise the compiler assumes no aliasing.